lib/Optimizer/GraphOptimizer/ConstantFolding.cpp

```diff
@@ -0,0 +1,219 @@
+bool isConstantOperation(const Node *N, const Backend &backend) {
+  // An operation with side-effects cannot be computed at compile-time.
+  if (N->hasSideEffects()) {
+    return false;
+  }
+  // Constant and splat nodes are trivially constant operations.
+  if (llvm::isa<Constant>(N) || llvm::isa<SplatNode>(N)) {
+    return true;
+  }
+  // If the operation is not supported by the backend, it cannot be computed at
+  // compile-time.
+  if (!backend.isOpSupported(NodeInfo(*N))) {
+    return false;
+  }
+  if (llvm::isa<Placeholder>(N)) {
+    return false;
+  }
+  for (unsigned idx = 0, e = N->getNumInputs(); idx < e; ++idx) {
+    auto input = N->getNthInput(idx);
+    if (!isConstantOperation(input.getNode(), backend)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/// Compile the function \p F for the provided \p backend using the compilation
+/// context \p cctx. \returns compiled function
+std::unique_ptr<CompiledFunction> compile(Backend &backend, Function &F,
+                                          CompilationContext &cctx) {
+  EXIT_ON_ERR(::glow::optimizeFunction(&F, backend, cctx));
+  for (const Node &N : F.getNodes()) {
+    CHECK(backend.isOpSupported(N))
+        << "Backend must support all nodes after high-level optimizations but "
+           "encountered unsupported operator: "
+        << N.getDebugDesc();
+  }
+  auto funcOrErr = backend.compile(&F, cctx.backendOpts);
+  EXIT_ON_ERR(funcOrErr.takeError());
+  return std::move(*funcOrErr);
+}
+
+/// Runs the compiled function \p compiledF from the module \p M on the \p
+/// backend using provided \p bindings.
+void run(Backend &backend, CompiledFunction &compiledF, Function &F,
+         PlaceholderBindings &bindings) {
+  std::unique_ptr<PlaceholderBindings> bindingsPtr(&bindings);
+  ExecutionContext context(std::move(bindingsPtr));
+  // Allocate only placeholders used by F.
+  for (auto ph : F.findPlaceholders()) {
+    // Don't allocate tensors for placeholders that are already allocated.
+    if (context.getPlaceholderBindings()->count(ph)) {
+      continue;
+    }
+    context.getPlaceholderBindings()->allocate(ph);
+  }
+  // TODO: Add only constants used by F to the compiled function. This should
+  // reduce the amount of data that needs to be copied.
+  auto executeErr = compiledF.execute(&context);
+  EXIT_ON_ERR(std::move(executeErr));
+  // don't delete bindings.
+  context.movePlaceholderBindings().release();
+}
+
+/// Execute function \p F by the \p backend using the provided \p bindings and
+/// the compilation context \p cctx.
+void execute(Backend &backend, Function &F, PlaceholderBindings &bindings,
+             CompilationContext &cctx) {
+  auto compiledF = compile(backend, F, cctx);
+  run(backend, *compiledF, F, bindings);
+}
+
+/// Evaluates a provided constant operation \p C in the context of the function
+/// \p F using the provided \p backend. Constant results are stored in \p
+/// constResults.
+void evaluateConstantOperation(
```

Comment:
nit: instead of `void` and passing `constResults`, could return `SmallVector`.

Reply:
I'll return an std::vector as it is simpler for the client code.